### ISSUE TITLE
Large tables performance

### ISSUE THREAD

**Issue by kayvanaarssen** [laravel-filament/filament#9304]
### Package

filament/filament

### Package Version

V3+

### Laravel Version

N/A

### PHP Version

PHP 8.2+

### Problem description

When using bigger tables and using the table size of 25+ for example 50 or All (Yeah All)
Than the loading times are huge.
The issue is also in the demo: https://demo.filamentphp.com/shop/orders

### Expected behavior

Since we have used other products and Vue.js / Laravel Backpack loading of tables was really quick with these kinds of datasets.
Would be awesome if we can see some huge improvements here.

The issue is most notable with 50+ items in a CRUD and sure All will be very heavy.
But a page size of 75 / 100 / 200 will sometimes with big datasets be a usecase.

### Steps to reproduce

If you select All here at the bottom 1000 records will be loaded. You have to wait a bit, and after that's done and you refresh.
The loading of the table will take about 14-18 seconds.
Does not matter how powerful the server is, and also your browser page may crash...

### Reproduction repository

https://github.com/filamentphp/filament

**Comment by kayvanaarssen** [comment]
It does not matter what project we link its already present in the base of Filament: https://github.com/filamentphp/filament

**Comment by NazarAli** [comment]
i have same problem 😞

**Comment by kayvanaarssen** [comment]
@zepfietje Can you maybe reopen this issue? Since its closed right after creation?
Thanks, and will schedule a meeting with you for next week :-D

**Comment by zepfietje** [comment]
Yeah fair, this is reproducible in the demo indeed.
Performance issues in large tables are known and likely caused by Livewire.
We've already tried to optimize as much as possible though.

Will reopen this issue, but we may close it again if there's nothing we can do.
Looping in @danharrin and @pxlrbt, since we've discusses this issue before.

If you have any ideas about approaches to improve performance, let us know, @kayvanaarssen!

**Comment by kayvanaarssen** [comment]
Thanks, talked about this already with our devs but no idea/ solution yet. Hope someone will be able to fix this / improve the speed.

**Comment by binaryfire** [comment]
I’m currently restricting the max results per page to 50 because of this. There are certain types of apps where viewing 100-200 records at a time would be desirable though. For example execution and monitoring logs in devops platforms.

**Comment by zepfietje** [comment]
I agree for certain apps it is important to view more than 100 records at once.
Running into this myself for [Picci](https://picci.app) too.

**Comment by kayvanaarssen** [comment]
But you think it has to do with livewire right?
And not Filament itself?
Maybe cross post in Livewire git as well?

**Comment by zepfietje** [comment]
Not sure if it really classifies as a Livewire *bug* though.
Posting in Livewire repo only really adds value when we create a failing test PR for bugs to be honest.

Maybe this is something @danharrin could discuss with Caleb though?

**Comment by kayvanaarssen** [comment]
Hope someone will find the issue / fix it. Coming from Backpack loading 500+ records was never an issue😉 Glad its not on our side only. But also in the base.

**Comment by binaryfire** [comment]
Hey @alecritson, I remember you mentioning wrangling with Livewire performance when creating Lunar’s datatables package. Maybe you have some insights you could share?

**Comment by kayvanaarssen** [comment]
Makes sense with ecom you wil have many tables with order/products etc. And filters will also be essential. Altough filters speed is okay in smaller tables, don't know how that will be with 200 or even 500 lines and lets say 30k of records.

**Comment by asefsoft** [comment]
Alright, I've conducted some debugging to identify the root cause. I will share my findings to determine if they can contribute to performance improvement or not. I cloned the demo project, seeded it with data, and loaded a page with 150 records on each page on the Order page.

Upon investigation, it became apparent that the issue arises when Livewire is rendering the Order Page component within the [update ](https://github.com/livewire/livewire/blob/f7d3918c507a8bbc994d2ce13bba56af15d04bfd/src/Mechanisms/HandleComponents/HandleComponents.php#L111) function. 

It takes more than **10 seconds** to fully render the component and generates approximately **3MB** of data!

Then I hooked into the `composing` event and saw that to render Order page it is rendering more that 5000 views!
I saw component names like this:

```
filament-tables::columns.text-column
filament-tables::components.cell
filament-tables::components.actions
filament-tables::components.actions.cell
filament-tables::components.cell
filament-tables::components.row
filament-tables::components.selection.checkbox
filament-tables::components.selection.cell
filament-tables::components.cell
filament-tables::components.columns.column
```
There are over **4000** `filament-tables*` rendering events, indicating a rendering process for each cell, row, and column.

you can review the entire rendering log that I have attached here: [compose_events-2023-11-10.log](https://github.com/filamentphp/filament/files/13318314/compose_events-2023-11-10.log)

I believe there might be some room for optimization in rendering tables, so I've decided to share this information with you for further investigation and potential improvement.

In conclusion, there is a significant number of rendering processes, and the HTML output appears to be very large, contributing to an extended rendering time for pages with 150 records.

**Comment by kayvanaarssen** [comment]
@zepfietje Is this something you guys can fix in Filament or is this more related to Livewire?
4000+ Filament events is a lot, like we saw also with debugging.

**Comment by pxlrbt** [comment]
> There are over 4000 filament-tables* rendering events, indicating a rendering process for each cell, row, and column.

I don't think that should be an issue. The views should be cached anyway.

> It takes more than 10 seconds to fully render the component and generates approximately 3MB of data!

3MB sounds like a lot of data and a huge table! I just compared this to the Filament Demo's "Brands" resource. With 400 (!) entries, it still sends only 127kB data and takes 2.3s to load for me.

It's hard to judge your "benchmark" without knowing your setup. If your table really generates 3MB of data, there is not much Filament can do.

Do you have Debugbar enabled?

**Comment by pxlrbt** [comment]
Just tested this again with the OrderResource. Demo with up-to-date Filament. 400 Orders load in 3.6s and produce 195kB in data (Macbook Air M1).

**Comment by kayvanaarssen** [comment]
> Just tested this again with the OrderResource. Demo with up-to-date Filament. 400 Orders load in 3.6s and produce 195kB in data (Macbook Air M1).

That's still a long wait time right for 400 lines... I mean, i get Filament has some overhead and cannot compare that to "Backpack" that we used to use, but over there 500+ Lines loaded within <1s

**Comment by asefsoft** [comment]
Hello Dennis 
I really dont know very much about internals of Filament and I just wanted to show that there is 4000 events and it might be problem or not. ofcource It need more investigation.

about taking 10 sec: it takes 10sec on my system with debug enable and i think it take half of it without it. I think anything above 2 sec or even 1 sec for 150 records is high.

and about size: in your demo site it produce 1.15MB for 50 records:
![image](https://github.com/filamentphp/filament/assets/46367425/cdb820c6-a480-4e8b-b609-8e6f8ed56f1e)

also in demo site it produce 20MB in 9 sec for 1000 records

![image](https://github.com/filamentphp/filament/assets/46367425/eabee35e-b913-41fc-8efd-e237755e482d)

**Comment by asefsoft** [comment]
The strange thing is that the 20MB is compressed into 162KB :))) you can see it in Transferred col.

**Comment by danharrin** [comment]
It's compressed because the HTML to render table cells is repetitive

If we were to reduce the number of views, then we would be introducing repetition within our codebase, since we wouldn't be able to use Blade components

**Comment by asefsoft** [comment]
> It's compressed because the HTML to render table cells is repetitive

Yes but strange part for me was the compression ratio which is over 99%, it became 125 times smaller. (wow)
It seems everything in it is repetitive, If my calculation was not wrong!

**Comment by danharrin** [comment]
Yes, Tailwind is quite repetitive 😉

**Comment by asefsoft** [comment]
Yes that's amazing and its impact is huge.
Okay, I want to delve deeper into the extensive response to see what I can find. Regardless of whether there is an easy way for optimization, I'll document it here. Perhaps I or someone else can come up with ideas and create a pull request for it, or share insights with our respected developers, Dan and Dennis.

I am currently examining a 150-record response weighing 3MB. After decoding it, I noticed some code snippets like this:

```
<!--[if ENDBLOCK]><![endif]--> 
<!--[if BLOCK]><![endif]-->
```

"It relates to [Livewire Morphing](https://livewire.laravel.com/docs/morphing#injecting-morph-markers) . Could we consider making an exception for tables and disabling it for them? You won't believe that there were more than **12,000** of each of them! By removing them, the response size reduced by about 22%

Afterward, I discovered some repeated classes in the output HTML that somehow might exceptionally could be compact for Tables or maybe not!

155 input with this classes:
`fi-checkbox-input rounded border-none bg-white shadow-sm ring-1 transition duration-75 checked:ring-0 focus:ring-2 focus:ring-offset-0 disabled:pointer-events-none disabled:bg-gray-50 disabled:text-gray-50 disabled:checked:bg-current disabled:checked:text-gray-400 dark:bg-white/5 dark:disabled:bg-transparent dark:disabled:checked:bg-gray-600 text-primary-600 ring-gray-950/10 focus:ring-primary-600 checked:focus:ring-primary-500/50 dark:text-primary-500 dark:ring-white/20 dark:checked:bg-primary-500 dark:focus:ring-primary-500 dark:checked:focus:ring-primary-400/50 dark:disabled:ring-white/10`

850 rows with this class
`fi-ta-text-item inline-flex items-center gap-1.5 text-sm text-gray-950 dark:text-white`

1000+:
`flex w-full disabled:pointer-events-none justify-start text-start`

1300+:
`fi-ta-cell p-0 first-of-type:ps-1 last-of-type:pe-1 sm:first-of-type:ps-3 sm:last-of-type:pe-3`

1000+:
`flex w-full disabled:pointer-events-none justify-start text-start`

150:
```
x-bind:class="{
            &#039;hidden&#039;: false &amp;&amp; isGroupCollapsed(&#039;&#039;),
            &#039;bg-gray-50 dark:bg-white/5&#039;: isRecordSelected(&#039;300&#039;),
            &#039;[&amp;&gt;*:first-child]:relative [&amp;&gt;*:first-child]:before:absolute [&amp;&gt;*:first-child]:before:start-0 [&amp;&gt;*:first-child]:before:inset-y-0 [&amp;&gt;*:first-child]:before:w-0.5 [&amp;&gt;*:first-child]:before:bg-primary-600 [&amp;&gt;*:first-child]:dark:before:bg-primary-500&#039;: isRecordSelected(&#039;300&#039;),
```

and there is more ...

**Comment by pxlrbt** [comment]
I don't think we should focus on the response size, since it's compressed anyway. It would be more interesting to see if we could improve rendering time by reducing classes etc. I don't think we can turn off Livewire internals for a specific part of the app.

**Comment by iprastha** [comment]
We have the same issue, tables with 500+ data always timed-out in nginx (504 gateway timeout), other pages with less data are quick.

**Comment by rickycheers** [comment]
TLDR; For anyone else running into this issue in dev mode, If you're using Laravel Debug Bar, try disabling it. (use this env variable `DEBUGBAR_ENABLED=false`)

For what is worth. I was having the same performance issue while attempting to render a table with barely 25 items per page but with more than 20 columns in it.

I noticed that the page would perform slightly better the less columns I added to the table. Then I started looking at the browser's profiler just to realize that the problem wasn't Filament or Livewire, per se... It was in fact caused by Larevel's debug bar.

Like someone else pointed out in earlier comments, many views get rendered in the back-end side when building the table. All of those rendering events are captured by the debug bar which causes the performance issue when it builds its UI.

**Comment by kayvanaarssen** [comment]
It also happens in Production and in the filament demo site. So than rhe debugbar is off.

**Comment by rickycheers** [comment]
@kayvanaarssen sorry, my reply wasn't directed specifically towards you, it is aiming anyone else stumbling upon this issue just to rule out that possibility. I'll update it to make it clearer.

**Comment by kayvanaarssen** [comment]
Understand, thought you were reffering to that the issue is only there when the debugbar is on😅

**Comment by andrewdwallo** [comment]
@danharrin @zepfietje I suggest deferring the loading of additional rows in a table until the user scrolls to the bottom of the viewport. This approach would enhance performance and user experience by loading more records only when necessary. I'm not sure if this has been considered as a solution, but it could be beneficial. Once the user reaches the bottom, more records can be loaded dynamically. However, I'm uncertain how this would impact existing pagination options. Perhaps this could be introduced as a separate feature for tables.

**Comment by andrewdwallo** [comment]
A similar solution instead of relying on the scroll position.
<img width="1529" alt="Screenshot 2024-05-22 at 1 50 49 PM" src="https://github.com/filamentphp/filament/assets/104294090/80c08c28-7a51-41d5-a6cf-6885f14e3fe5">

**Comment by zepfietje** [comment]
Would that really make a difference? Because I think DOM diffing is the main cause of performance issues?

**Comment by danharrin** [comment]
The real solution for rendering tables faster is probably using Alpine.js to render cells in a JS loop from JSON data instead of with Blade, but its a huge breaking change and would also limit the stuff you could do in a cell as you wouldn't have full access to Blade

**Comment by andrewdwallo** [comment]
> Would that really make a difference? Because I think DOM diffing is the main cause of performance issues?

I understand the concern about DOM diffing potentially being the main cause of performance issues. However, I'd like to clarify the benefits of deferring the loading of rows until the user scrolls to the bottom (infinite scrolling or lazy loading):

1. **Reduced Initial Load Time:** By loading only a small subset of rows initially, we significantly reduce the amount of data rendered at once. This should improve the initial load time and make the page more responsive.
2. **Incremental Data Loading:** As users scroll, additional rows are loaded in smaller batches. This can help manage memory usage and render time more effectively, reducing the risk of browser crashes with large datasets.
3. **Improved User Experience:** Users cannot currently view hundreds of records at once in the current viewport (only around 10-18 at once at 100% zoom on a 1920x1080 resolution), so there is no reason to load 25/50/100+ or all records at once. Loading data incrementally as they scroll can provide a smoother and faster experience. This ensures that the visible records load quickly without overwhelming the browser.

Regarding the impact on existing pagination options, infinite scrolling can be implemented alongside traditional pagination seamlessly. For example, when users choose to view 25/50/100+ records per page or all records, my suggestion can take action. This allows both methods to coexist effectively, enhancing performance while maintaining user preference for pagination.

I understand that integrating this feature might require significant changes and careful consideration of its impact on other functionalities. It may also be beneficial to explore combining this approach with Alpine.js or other frontend optimizations as mentioned.

Would you be open to exploring this further as a potential enhancement? I’m happy to collaborate on this and help test any solutions we come up with.

**Comment by andrewdwallo** [comment]
Based on the performance testing I conducted, I'd like to address some additional points and suggestions:

### Metrics Summary

#### 10 records per page

```
Load timings (ms)
```
| Event            | Start | Duration | End  |
|------------------|-------|----------|------|
| Redirect         | 0     | 0        | 0    |
| DNS              | 0     | 0        | 0    |
| Connect          | 0     | 0        | 0    |
| Request          | 3     | 367      | 370  |
| Response         | 370   | 1        | 371  |
| DOM              | 371   | 135      | 506  |
| Parse            | 371   | 37       | 408  |
| Execute Scripts  | 408   | 13       | 421  |
| Content loaded   | 421   | 20       | 441  |
| Sub Resources    | 441   | 65       | 506  |
| Load event       | 506   | 1        | 507  |
| **Total**        |       |          | 507  |

#### 25 records per page

```
Load timings (ms)
```
| Event            | Start | Duration | End  |
|------------------|-------|----------|------|
| Redirect         | 0     | 0        | 0    |
| DNS              | 1     | 0        | 1    |
| Connect          | 1     | 0        | 1    |
| Request          | 8     | 563      | 571  |
| Response         | 571   | 0        | 571  |
| DOM              | 571   | 162      | 733  |
| Parse            | 571   | 67       | 638  |
| Execute Scripts  | 638   | 1        | 639  |
| Content loaded   | 639   | 15       | 654  |
| Sub Resources    | 654   | 79       | 733  |
| Load event       | 733   | 1        | 734  |
| **Total**        |       |          | 734  |

#### 50 records per page

```
Load timings (ms)
```
| Event            | Start | Duration | End  |
|------------------|-------|----------|------|
| Redirect         | 0     | 0        | 0    |
| DNS              | 1     | 0        | 1    |
| Connect          | 1     | 0        | 1    |
| Request          | 3     | 930      | 933  |
| Response         | 933   | 0        | 933  |
| DOM              | 933   | 192      | 1125 |
| Parse            | 933   | 61       | 994  |
| Execute Scripts  | 994   | 0        | 994  |
| Content loaded   | 994   | 16       | 1010 |
| Sub Resources    | 1010  | 115      | 1125 |
| Load event       | 1125  | 1        | 1126 |
| **Total**        |       |          | 1126 |

#### 100 records per page

```
Load timings (ms)
```
| Event            | Start | Duration | End  |
|------------------|-------|----------|------|
| Redirect         | 0     | 0        | 0    |
| DNS              | 0     | 0        | 0    |
| Connect          | 0     | 0        | 0    |
| Request          | 3     | 1258     | 1261 |
| Response         | 1261  | 0        | 1261 |
| DOM              | 1261  | 252      | 1513 |
| Parse            | 1261  | 93       | 1354 |
| Execute Scripts  | 1354  | 0        | 1354 |
| Content loaded   | 1354  | 19       | 1373 |
| Sub Resources    | 1373  | 140      | 1513 |
| Load event       | 1513  | 1        | 1514 |
| **Total**        |       |          | 1514 |


### Key Findings

1. **Request Duration:** 

- The time taken for the request to complete increases significantly with the number of records, suggesting that server-side processing and data retrieval times grow proportionally with the number of records.

2. **DOM Processing:** 

- The DOM processing time also increases with the number of records, indicating that rendering the additional records in the browser takes more time as the number of records grows.

3. **Sub Resource Loading:** 

- The time taken to load sub resources (like CSS, JavaScript, images) also increases, though to a lesser extent.

### Relevance to DOM Diffing and Alpine.js Suggestion

While the primary bottleneck appears to be server-side processing, the increase in DOM processing time suggests that the efficiency of DOM updates (including DOM diffing) is also a factor.

**Regarding DOM Diffing:**

- DOM diffing helps optimize updates to the DOM by only re-rendering elements that have changed. However, the increased DOM processing time with more records indicates that the sheer volume of data being rendered might still overwhelm this optimization.

**Considering Alpine.js:**

- Alpine.js could be used to render cells in a JavaScript loop from JSON data instead of using Blade. This approach could improve client-side performance by optimizing DOM updates and reducing the overhead of server-side Blade rendering.
- However, as mentioned by Dan, this would require a significant refactor of the existing codebase and could limit the flexibility provided by Blade, such as using Blade directives and custom server-side logic directly within the view.

### Recommendations

1. **Optimize Server-Side Processing:**
- Review and optimize database queries and server-side logic to handle large datasets more efficiently.

2. **Implement Lazy Loading/Infinite Scrolling:**
- Loading data incrementally can significantly reduce the initial load time and improve overall performance by distributing the load over time as the user scrolls. This can be implemented alongside traditional pagination, activating when users choose to view larger numbers of records per page.

5. **Efficient Client-Side Rendering:**
- Explore ways to reduce the complexity of the DOM generated for large datasets, such as using virtual DOM techniques or frontend frameworks like Alpine.js.

6. **Leverage Caching:**
- Implement caching strategies both on the server-side and client-side to reduce redundant processing and data retrieval.

### Conclusion

The suggestion to use Alpine.js to render cells from JSON data is a valid long-term solution to improve rendering performance. However, as Dan mentioned, it would require careful consideration of the trade-offs, including the significant refactor required and the loss of Blade's flexibility. Combining this approach with server-side optimizations and lazy loading could provide a balanced solution to improve performance.

**Comment by wychoong** [comment]
Scroll loading and alpinejs seems increase the complexity. 
why not consider the new wire:replace https://github.com/livewire/livewire/pull/8403

**Comment by mentamarindos** [comment]
You guys as @wychoong mentions (Livewire v3.5.0) Adding **wire:replace** by @samlev in https://github.com/livewire/livewire/pull/8403  
may solve this issue

**Comment by kl4ver** [comment]
> You guys as @wychoong mentions (Livewire v3.5.0) Adding **wire:replace** by @samlev in [livewire/livewire#8403](https://github.com/livewire/livewire/pull/8403) may solve this issue

Are there any plans to add this? I see it's got the label low priority, but I think this is for a lot of users a big issue.

**Comment by wychoong** [comment]
> > You guys as @wychoong mentions (Livewire v3.5.0) Adding **wire:replace** by @samlev in [livewire/livewire#8403](https://github.com/livewire/livewire/pull/8403) may solve this issue
> 
> Are there any plans to add this? I see it's got the label low priority, but I think this is for a lot of users a big issue.

It’s open source project and filament is provided for free. For **“a lot of users” and “big issue”**, there seems to be not enough interest to provide a repo with huge data slow performance table, benchmarking setup, let alone funding the issue or submit a PR. 

Regarding my suggestion of wire:replace, one should understand that it’s a relatively new api in LW and it might be optional enable in v3 or only available in v4

**Comment by kl4ver** [comment]
> > > You guys as @wychoong mentions (Livewire v3.5.0) Adding **wire:replace** by @samlev in [livewire/livewire#8403](https://github.com/livewire/livewire/pull/8403) may solve this issue
> > 
> > 
> > Are there any plans to add this? I see it's got the label low priority, but I think this is for a lot of users a big issue.
> 
> It’s open source project and filament is provided for free. For **“a lot of users” and “big issue”**, there seems to be not enough interest to provide a repo with huge data slow performance table, benchmarking setup, let alone funding the issue or submit a PR.
> 
> Regarding my suggestion of wire:replace, one should understand that it’s a relatively new api in LW and it might be optional enable in v3 or only available in v4

Thanks for the explanation, I thought I would be a small change.
What you say about provide a demo, it's also a problem on https://demo.filamentphp.com/blog/categories, if you set the pagination to all (only 400 records), the page takes 4 seconds to load.

**Comment by danharrin** [comment]
The amount of HTML that needs to be loaded for those 400 records is substantial enough to slow the page down. In most projects, I would remove the "All" option and limit it to 50 per page maximum.

**Comment by mihob** [comment]
To be honest, it is quite slow even with 50 entries.  400-500ms is pretty much for a table with a few entries.

The problem does not only apply to tables, but also more or less runs through entire filament.

For example, if I have a form with a repeater field, the request time also increases unreasonably with each entry

**Comment by kayvanaarssen** [comment]
> To be honest, it is quite slow even with 50 entries. 400-500ms is pretty much for a table with a few entries.
> 
> The problem does not only apply to tables, but also more or less runs through the entire filament.
> 
> For example, if I have a form with a repeater field, the request time also increases unreasonably with each entry

Totally agree on this. We experience the same symptoms.

**Comment by danharrin** [comment]
The more items you add to a repeater field, the more HTML is required to render that repeater

**Comment by mihob** [comment]
I am aware of that. But the difference with a few entries is quickly in the region of several 100ms.

If you were to render this with a few blade components, it would be much much faster.

Don't get me wrong. I really like filament, but I think performance is a problem that you can't just dismiss.

You very quickly reach request times of over 1s and you almost can't expect customers to accept that.

**Comment by danharrin** [comment]
We're working on optimisations involving the HTML size and partial rendering for version 4, but there are some limitations that cannot be overcome without compromising in other areas

**Comment by shivaRamdeen** [comment]
html is over 14mb. Is this expected?
This is loading 200 rows on a table most columns are hidden.
![image](https://github.com/filamentphp/filament/assets/14797072/2823a313-597b-4c65-83dc-159d4b9275da)

**Comment by mgkimsal** [comment]
> html is over 14mb. Is this expected? This is loading 200 rows on a table most columns are hidden. 

Even if the columns are hidden, I believe they're still calculated and returned down, no?  Maybe I'm wrong on that one.

**Comment by danharrin** [comment]
They are not

**Comment by shivaRamdeen** [comment]
@danharrin do you have any pointers on resolving this issue? I am willing to dedicate the time to develop a solution as this is a major issue.

**Comment by danharrin** [comment]
There isn't one "resolution", just a series of investigations and small optimisations that can be made from your findings.

**Comment by binaryfire** [comment]
> The real solution for rendering tables faster is probably using Alpine.js to render cells in a JS loop from JSON data instead of with Blade, but its a huge breaking change and would also limit the stuff you could do in a cell as you wouldn't have full access to Blade

Are there any internal discussions about taking this approach? Personally I'd be fine with losing some functionality in return for client-side datatable performance. Might be worth doing a poll? I've never seen anyone complain about Nova or Backpack's datatables (or any of the popular Vue or React datatabes packages).

**Comment by pxlrbt** [comment]
> Personally I'd be fine with losing some functionality in return for client-side datatable performance

This is probably a *huge* breaking change. For 3 versions we fully rely on Livewire for all the logic and users can define their own conditions, interactions between actions and the components, etc.

**Comment by shivaRamdeen** [comment]
Proposed path towards client side support without breaking existing tables:

Fully client side tables.

- [ ] Add ->isClientSide() to existing tables. Renders a table view that makes API requests to get the table schema and data.
- [ ] Build read only cells renderable with js/alpine based on filament table schema.
- [ ] Investigation: Editable cells need a protocol to interact with the filament/livewire table to CRUD cell values.
- [ ] Build editable cells using defined protocol.

**Comment by danharrin** [comment]
And who is funding the maintenance of this new table builder 😁 I don't think it's practical to maintain both for us

**Comment by shivaRamdeen** [comment]
If funding is a problem, what about making it a third party plugin instead?

**Comment by danharrin** [comment]
It's just a lot of work to maintain, whoever does it, I am going to wait and see how much improvement we can make from the changes in v4

**Comment by pxlrbt** [comment]
> If funding is a problem, what about making it a third party plugin instead

Well, go ahead 😅

**Comment by zepfietje** [comment]
I just spent a considerable amount of time debugging and profiling the table performance issue using the orders list page in the demo and showing 100 records at once.

It seems the issue isn't related to Livewire at all, as initial page load times are almost identical to subsequent requests when visiting the next page of table records (~800ms).

A couple of potential solutions I've tried:

### Removing CSS classes

Removing CSS classes from all components rendered in the table didn't make any significant difference in loading times.

### Replacing Blade components

Replacing Blade components like table rows with plain HTML elements made somewhat of a difference, but not enough to solve the performance issue.

### Simplifying Blade components

The text column Blade component currently contains a lot of nested HTML elements which aren't necessary when displaying simple text values. Removing these elements resulted in a slightly noticeable performance improvement (~800ms down to ~500ms).

Unfortunately it doesn't seem like any of these approaches can fully solve the issue.

**Comment by andrewdwallo** [comment]
@zepfietje I find it hard to believe that loading 100 rows only takes 800ms. Is this version of the table you're using very light and not complex? How many columns? Could the performance issue simply be the amount of conditions being checked  while rendering a table? I appreciate the time you took debugging.

Edit: Also, of course loading all rows for a table would take a long time anyway. Filament couldn't really do anything about that other than using infinite scroll to reduce the amount of html being rendered at once. You wouldn’t need to incrementally load records from the database; you could simply use the existing dataset while loading the HTML in chunks.

**Comment by andrewdwallo** [comment]
> Edit: Also, of course loading all rows for a table would take a long time anyway. Filament couldn't really do anything about that other than using infinite scroll to reduce the amount of html being rendered at once. You wouldn’t need to incrementally load records from the database; you could simply use the existing dataset while loading the HTML in chunks. 

If you think this would be a good idea I can try to implement something over the weekend and get back to you.

**Comment by malle-pietje** [comment]
> > Edit: Also, of course loading all rows for a table would take a long time anyway. Filament couldn't really do anything about that other than using infinite scroll to reduce the amount of html being rendered at once. You wouldn’t need to incrementally load records from the database; you could simply use the existing dataset while loading the HTML in chunks.
> 
> If you think this would be a good idea I can try to implement something over the weekend and get back to you.

From my experience in working with the jQuery Datatables package a lot I can tell you that it easily becomes unworkable when you're loading 20000 rows or more; the browser either chokes or your PHP process runs out of memory (I'd rather not assign more than 1GB of memory to my PHP processes).

The main benefit of Datatables is that it is blazingly fast once the data has been loaded 😉

**Comment by rvzug** [comment]
> ### Replacing Blade components
> Replacing Blade components like table rows with plain HTML elements made somewhat of a difference, but not enough to solve the performance issue.
> 
> ### Simplifying Blade components
> The text column Blade component currently contains a lot of nested HTML elements which aren't necessary when displaying simple text values. Removing these elements resulted in a slightly noticeable performance improvement (~800ms down to ~500ms).

I also dug into a profiling session and the rendering of each table cell adds up to the total response time. Isn't it possible to somehow cache the table row before rendering, and then render each row from cache? Or parts of it, like table actions with icons and/or nested components. It would at least reduce the IO-operations on the disk.

I also noticed that large tables are killing responses on AWS Lambda, using Laravel Vapor. Resulting in a 502 which probably means you ran into a Lambda quotation.

**Using Vapor**
In a project using Vapor, it did help a lot to change the compiled view-path to ```/tmp/views``` (as other directories are not writable):

config/view.php change:
```php
    'compiled' => '/tmp/views',
```

vapor.yml
```yml
environments:
      env:
          build:
              - 'php artisan optimize'
              - 'php artisan filament:optimize'
```

**Comment by pxlrbt** [comment]
> Isn't it possible to somehow cache the table row before rendering

I don't think that works, because you can use `fn ($record) => ` for most of the methods on columns, so each row and column must be reevaluated. We can't simple save the HTML and replace stuff.

**Comment by danharrin** [comment]
I do have an idea about how we can refactor tables to render using an Alpine loop without breaking anything, but I haven't tested it and it would be a lot of work. And I don't know if it would necessarily provide much of a performance improvement (it would push some work from the backend into the frontend). Could be worth attempting though.

**Comment by zepfietje** [comment]
> I do have an idea about how we can refactor tables to render using an Alpine loop without breaking anything, but I haven't tested it and it would be a lot of work. And I don't know if it would necessarily provide much of a performance improvement (it would push some work from the backend into the frontend). Could be worth attempting though.

Might be worth experimenting with, though from my experience using Alpine on large parts of a page results in performance issues too (make sure to check all browsers as this was especially bad in Safari IIRC).

**Comment by galliroberto** [comment]
The solution is to move to Svelte :)

**Comment by pmartelletti** [comment]
Wondering if we've considered what Philos discussed in [this video](https://www.youtube.com/watch?v=857ean0JIHE&ab_channel=Laravel)?

I know Filament renders thousands of blade components each time something changes in the table. One of the big performance issues he discussed is changing that (which is essentially HTML) into livewire components, as livewire is clever enough just to change the livewire component that changed, as opposed to doing diffing with all the document.

Is it something you think it's worth exploring?

**Comment by danharrin** [comment]
The answer is no, in our situation it would not help. The overhead of booting so many more Livewire components would be more than we have now, and there would be a significant delay in communicating updates through events once something changes.

Once you start breaking up Livewire components, you cannot render them all at once again and just choose which ones need to change. Livewire is just assuming each component is an island that doesn't know anything about the others.

**Comment by pmartelletti** [comment]
Thanks for the clarification @danharrin  :)

**Comment by NazarAli** [comment]
i use wamp server php 8.3  is there any config can speed filament table performance ?

**Comment by nocodelab** [comment]
Same performance issues here. Large table are a pain and I have tried everything (without success) to improve the performances.
A Livewire only approach makes the development super easy and reactive but for medium-large applications there is a big gap compared to JS DataTable performances.

**Comment by shivaRamdeen** [comment]
@nocodelab For performance writing custom widgets seems like the best available option unfortunately.

you can reuse filament components in your widget code and write alinejs to handle everything in JavaScript. Much more performant.

yes, I know rewriting something that already exists is a pain but no other solution atm.

**Comment by miagg** [comment]
Same issue here. On a table with just 8 rows I get 250ms loading time even with filament cache enabled.
According to Sentry 100ms is being spent on rendering `filament-tables::components.actions` component (~10ms per table action)
I can confirm that because once I remove all table actions, loading is dropped to ~140ms.

I guess nesting views takes considerable time to render even when cached so this does not scale well. 
I hope Filament 4 will try to solve this, otherwise it has no feature. At least for large-scale apps 😢

<img width="1019" alt="perf1" src="https://github.com/user-attachments/assets/c0c88c17-eaf1-4dc9-9b02-59721e250f5e">
<img width="1127" alt="perf2" src="https://github.com/user-attachments/assets/12d0c889-14c3-4b2c-aa0c-87c8f0cbab3a">

**Comment by danharrin** [comment]
This has improved in version 4. I do not see any way to port those fixes to v3 without introducing breaking changes, so I am closing this thread. I hope you like the improvements when that new version is released. And no, as of today, I don't have an ETA for the release.

**Comment by fero8** [comment]
I am facing the same issue, @danharrin would it be possible to point me in the right direction on how this has been addresses in v4 please?

**Comment by pxlrbt** [comment]
> Rendering and interaction performance in Filament has significantly improved, especially for large tables. Internally, many Blade templates have been optimized to reduce the number of views that are being rendered. By utilizing existing PHP objects to render HTML instead of including new files, Filament v4 reduces the number of files that need to be loaded, improving performance. The size of Blade views has also been reduced by extracting groups of Tailwind CSS classes into dedicated classes, which are then used in the Blade templates. This reduces the amount of HTML that needs to be rendered, resulting in faster page loads and a smaller response size.

https://filamentphp.com/content/leandrocfe-filament-v4-beta-feature-overview#performance

**Comment by fero8** [comment]
Just a quick update after using the new v4 for about a week. Yes, performance has improved quite a bit, w/out any code enhancements, just by upgrading. Even without, the new JS optimization methods. Nice job!
